### PR TITLE
Http request middleware function

### PR DESCRIPTION
**Pull request by functionstackx**
## Why
In order to propagate trace context & headers for distributed tracing, we need some way of "injecting" middleware to our requests

## How?
- an parameter was added to the the function that creates a new webrpc client

## Tests
- re-generated `_examples/golang-basic/generate.gen.go`
- `webrpc-gen -schema ./_examples/golang-basics/example.ridl -target go -out ./_examples/golang-basics/example.gen.go -client -server -pkg main`

- re-generated in internal repos

## Communication
- post on slack that if anybody needs makes modifications to internal private repo's `ridl`, they should `git pull`, re-build, and reinstalling webrpc

## Example Generated Client - `_examples/golang-basics`
image

## All Generated Client Helpers function will look like this
image
 
## Type of Function Pointer to Pass as a parameter into Create New Client
image

### CODE REVIEW

File: gen/golang/templates/client.go.tmpl [hunk 1]
```diff
@@ -21,20 +23,30 @@ const {{.Name | constPathPrefix}} = "/rpc/{{.Name}}/"
 {{range .Services}}
   {{ $serviceName := .Name | clientServiceName}}
   type {{$serviceName}} struct {
-    client HTTPClient
-    urls   [{{.Methods | countMethods}}]string
+    client         HTTPClient
+    urls           [{{.Methods | countMethods}}]string
+    InjectFunction InjectHTTPRequestFunction
```
Comment (kipply) [inline gen/golang/templates/client.go.tmpl]: can and probably should avoid the `...` and `[0]`?

Comment (functionstackx) [reply gen/golang/templates/client.go.tmpl]: do you want it to be a mandatory parameter instead of optional?

**Review by kipply — APPROVED** [review]

File: gen/golang/templates/client.go.tmpl [hunk 2]
```diff
@@ -4,6 +4,8 @@
 // Client
 //
 
+type InjectHTTPRequestFunction func(*http.Request, context.Context) *http.Request
```
Comment (miguelvr) [inline gen/golang/templates/client.go.tmpl]: rename to `InjectHTTPRequestFunc` (usually the Func suffix is the convention)

**Review by miguelvr — APPROVED** [review]

**Comment by VojtechVitek** [comment]
Hi @OrenLeung @miguelvr @kipply,

I'm walking through this fork's Pull Requests and wondering what we could backport to upstream [webrpc repo](https://github.com/webrpc/webrpc).

Can you please share how this feature is used in a real world example?
Are you setting some request headers based on the context values?

Example from https://rakyll.medium.com/context-propagation-over-http-in-go-d4540996e9b0:

```go
const requestIDHeader = "request-id"

// Transport serializes the request context into request headers.
type Transport struct {
 // Base is the actual round tripper that makes the request.
 // http.DefaultTransport is used if none is set.
 Base http.RoundTripper
}

// RoundTrip converts request context into headers
// and makes the request.
func (t *Transport) RoundTrip(r *http.Request) (*http.Response, error) {
 r = cloneReq(r) // per RoundTrip interface enforces
 
 rid := request.IDFromContext(r.Context())
 if rid != "" {
  r.Header.Add(requestIDHeader, rid)
 }
 
 base := t.Base
 if base == nil {
  base = http.DefaultTransport
 }
 
 return base.RoundTrip(r)
}
```

Have you considered using custom `http.Transport` instead? Would it be sufficient?

Thanks!

Best,
Vojtech | https://golang.cz

**Pull request merged**